### ISSUE TITLE
Generated schema depends on platform

### ISSUE THREAD

**Issue by Ten0** [diesel-rs/diesel#2202]
### Versions

- **Rust:** Latest stable
- **Diesel:** CLI from cargo
- **Database:** Postgres
- **Operating System:** Linux/MacOS (schema differs depending which one we're on)

### What are you trying to accomplish?
Reliably generate a `schema.rs` on whichever platform based on a set of migrations.
We are then patching the schema using `diff -u`/`patch`. The order being different breaks the patch, as well as deploy-time checks that ensure the git-tracked file is consistent with the database schema.

### What is the expected output?
On both platforms, `schema.rs` lines are ordered the same way.

### What is the actual output?
The tables are not ordered in the same way.
On Linux:
- d_c_e
- d_cs
- ds
- d_s_c

On Mac:
- d_c_e
- d_cs
- d_s_c
- ds

(Can be noted that the Mac ordering is more consistent. Looks like the Linux sort gets rid of the underscores before sorting.)

### Any clue on how to solve?

I suspect there are no `sort`s on the rust side when generating the schema, and this inconsistency may come more from the postgres side than from the rust side.
(Everytime schema generation was run on Mac so far, the database was also there. I may soon check for what happens when generating on MacOS a schema for a database that lives on a Linux postgres.)

It is possible that the fix is as simple as applying a sort on the rust side.
To limit breakage, the sort applied could be the same as is applied on the Linux version.

**Comment by weiznich** [comment]
Sorting the table names is [done on sql side](https://github.com/diesel-rs/diesel/blob/8396134f1fb581529070de0b2a946f1c7d5f9e91/diesel_cli/src/infer_schema_internals/information_schema.rs#L190) for postgresql.

As additional step to see if that's really the issue please execute the following sql query on both databases, to see if the returned order is equal or not.
```sql
SELECT table_name, table_schema from information_schema.tables 
WHERE table_schema = 'your schema' AND table_name NOT LIKE '__%'  AND table_type LIKE 'BASE TABLE' ORDER BY table_name;
```
If those results are not equal I think we should consider opening a bug at the postgresql issue tracker. In that case we could do the othering at the linked place on rust side as workaround. Otherwise we need to find the real issue.

**Comment by Ten0** [comment]
Hello,

We've run the query on a Linux server and on a Mac server, and we can confirm that these queries have different outputs: the Linux server seems to get rid of the underscores before sorting, while the Mac version doesn't.

**Comment by Ten0** [comment]
Reported on Postgres' bugs mailing list: https://www.postgresql.org/message-id/16077-6f297cabd744c86c@postgresql.org

**Comment by Ten0** [comment]
And there's been an answer:
https://www.postgresql.org/message-id/11505.1571936496%40sss.pgh.pa.us

**Comment by Ten0** [comment]
So it would seem we need to sort on our side.

**Comment by ralpha** [comment]
It looks like this should have been fixed in #2213 but I'm still having this problem with `diesel_cli` v1.4.1.
I'm using Postgres 11.13 and 12.8 (same version of `diesel_cli`)

In PG 11.13 it sorts:
```
a_b
a_r
as
a_st
```
In PG 12.8 it sorts:
```
a_b
a_r
a_st
as
```
Both are tested on Linux (same machine, just different databases)

Thus it seems to ignore the `_` characters in some way.
Should I open a new issue for this? Or should this issue be reopened?

**Comment by weiznich** [comment]
@ralpha This is expected because this commit is only available in the master branch, not in any release branch. Its scheduled to be released with the 2.0 release.

**Comment by ralpha** [comment]
> @ralpha This is expected because this commit is only available in the master branch, not in any release branch. Its scheduled to be released with the 2.0 release.

Oh okay, I see. It is not a major issue so I can live with it.
It would be nice to see v2.0 release, but I guess patience is a good thing :smile:

**Comment by Ten0** [comment]
I suspect if you were to
```bash
cargo install --git https://github.com/diesel-rs/diesel.git --rev 7c0e9fb00cd8fe43690bb82af19afb15ac518981 diesel_cli
```
you'd get a version that has this fixed while still being compatible with diesel v1.4.
